sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataTimeBasedBufferingOutboundObserver.java

```diff
@@ -90,20 +90,27 @@ private void periodicFlush() {
     try {
       flush();
     } catch (Throwable t) {
-      if (t instanceof IOException) {
-        flushException.set((IOException) t);
-      } else {
-        flushException.set(new IOException(t));
-      }
       throw new RuntimeException(t);
     }
   }
 
   /** Check if the flush thread failed with an exception. */
   private void checkFlushThreadException() throws IOException {
-    IOException e = flushException.get();
-    if (e != null) {
-      throw e;
+    if (flushFuture.isDone()) {
+      try {
+        flushFuture.get();
+        throw new IOException("Periodic flushing thread finished unexpectedly.");
+      } catch (ExecutionException ee) {
+        // the cause of ExecutionException is always RuntimeException
+        RuntimeException re = (RuntimeException) ee.getCause();
+        if (re.getCause() instanceof IOException) {
+          throw (IOException) re.getCause();
+        }
+        throw re;
+      } catch (InterruptedException | CancellationException e) {
+        // Should never happen
```

Comment:
Prefer to wrap the InerruptedException and CancellationException in an IOException.

Also for the interrupted exception you should invoke Thread.interrupt() before rethrowing.